component/cache.js

```diff
@@ -0,0 +1,40 @@
+var assert = require('assert')
+var LRU = require('nanolru')
+
+module.exports = ChooComponentCache
+
+function ChooComponentCache (state, emit, lru) {
+  assert.ok(this instanceof ChooComponentCache, 'ChooComponentCache should be created with `new`')
+
+  assert.equal(typeof state, 'object', 'ChooComponentCache: state should be type object')
+  assert.equal(typeof emit, 'function', 'ChooComponentCache: state should be type function')
+
+  this.cache = lru || new LRU(100)
```

Comment:
An LRU cache would start evicting components that are still on the page if you have a lot of them (eg. 120). I think components that are still mounted should always stay in the cache. Maybe we can have two caches, one plain object `.mounted`, and one LRU (or user provided) `.unmounted`. Then we can use on-load to move components into the unlimited `.mounted` cache on `load`, and move them to the bounded `.unmounted` cache on `unload`.

Reply:
simpler is probably better, and i like your suggestion @goto-bus-stop, but here's one other idea: i wonder if it makes sense to have separate caches per component class? i guess it would just be nice not to thrash header/footer/button singleton components if an app happens to have like 1000 tweet component instances...

Reply:
That would help too! Id still be worried about the tweets though. as I understand it a list of 100+ tweets would end up recreating and evicting instances during every render even if their content didn't change...

Reply:
The onload idea might be pretty difficult to implement in a generic way, though, I'm not sure.

Reply:
> Id still be worried about the tweets though

If people are thinking of creating lots and lots of elements, we could recommend they create a separate cache for this specifically. If we expose the cache instance, they'll have the flexibility of choosing elements and such.

Reply:
+1 on what @yoshuawuyts said but to make that easier couldn't we allow for the third argument (`lru`) to be a number and forward that to `nanocache`? So if you don't have any preference as to what lru cache to use, but just want to allow for a lot of instances you'd do `var manyTweetsCache = new ComponentCache(state, emit, 250)`

Also, most classes in the choo stack allows you to create instances without `new` using an `instanceof` check and calling `new` for you if needed. To maintain a uniform api, shouldn't we do that here as well?